### PR TITLE
enable filters in list_all_** apis

### PR DESCRIPTION
**Pull request by mshaban-msft**
solves #21213 
still not completed
just wanting to get some insights from reviewers!

### CODE REVIEW

File: sdk/translation/Azure.AI.Translation.Document/src/DocumentFilter.cs [hunk 1]
```diff
@@ -1,16 +1,42 @@
...
-    /// <see cref="TranslationSource"/> class.
+    /// A class which defines filtering and ordering options
+    /// for listing all document statuses for a certain translation operation.
     /// </summary>
-    [CodeGenModel("DocumentFilter")]
     public partial class DocumentFilter
     {
+        /// <summary>
+        /// Filter results by <see cref="DocumentStatus.CreatedOn"/>.
+        /// Get documents created AFTER a certain datetime.
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/src/DocumentFilter.cs:18]: same docstring comments for `TranslationFilter`

File: sdk/translation/Azure.AI.Translation.Document/src/DocumentFilterOrder.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        /// <summary>
+        /// Initializes an instance of <see cref="DocumentFilterOrder"/>.
+        /// </summary>
+        public DocumentFilterOrder(DocumentFilterProperty property, DocumentTranslationFilterMode orderBy)
+        {
+            Property = property;
+            OrderBy = orderBy;
+        }
+        /// <summary>
+        /// The order by which we sort results.
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/src/DocumentFilterOrder.cs:25]: Same comments for `TranslationFilterOrder`

File: sdk/translation/Azure.AI.Translation.Document/src/TranslationFilterOrder.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        /// <summary>
+        /// sorting property corresponds to <see cref="TranslationStatus.CreatedOn"/>.
+        /// </summary>
+        CreatedOn = 0,
+    }
+
+    /// <summary>
+    /// An enum listing the order by which we sort results.
+    /// </summary>
+    public enum DocumentTranslationFilterMode
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/src/TranslationFilterOrder.cs:54]: Could you move this to its own file, please?

File: sdk/translation/Azure.AI.Translation.Document/src/TranslationFilterOrder.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        /// <summary>
+        /// sort ascendingly.
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/src/TranslationFilterOrder.cs:57]: Start with upper case. applies to multiple docstrings

**Review by maririos — COMMENTED** [review]
The general implementation looks good!
Most comments are around:
- Improving docstrings (start with Upper case, use full words instead of `info`, etc.
- .NET specifics

Other comments:
- We could either add mock tests for all filters to make sure we do the serialization correctly or we can add a live test per type of filter to make sure we get what we expect. We don't need to be suuuper implicit about it, but at list see that the lists are properly serialized, we expose the date correctly, and that we are doing a good `orderBy` transformation`.
- Need to add recordings
- Samples: Do you think we should add a sample with filters so we can show the customer this exists? I don't think it is fully necessary but curious about what you think.

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        [Ignore("no way to test this filter")]
+        [RecordedTest]
+        public void GetAllDocumentStatusesFilterByCreatedAfter()
+        {
+        }
+
+        [Ignore("no way to test this filter")]
+        [RecordedTest]
+        public void GetAllDocumentStatusesFilterByCreatedBefore()
+        {
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs:80]: you could create mock tests for this and just verify that in the SDK we are sending the correct parameter name and date format.
An example of a mock test in TA is: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/textanalytics/Azure.AI.TextAnalytics/tests/AnalyzeOperationMockTests.cs#L42

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+        [RecordedTest]
+        public async Task GetAllTranslationStatusesFilterByCreatedAfterTest()
+        {
+            // create client
+            var client = GetClient();
+
+            // create test jobs
+            await CreateTranslationJobsAsync(client, jobsCount: 5, docsPerJob: 1, jobTerminalStatus: DocumentTranslationStatus.Succeeded);
+            var timestamp = DateTimeOffset.UtcNow;
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs:102]: you should use `Recording.UtcNow;` to get consistency in recordings

File: sdk/translation/Azure.AI.Translation.Document/src/DocumentFilterOrder.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        /// <summary>
+        /// The <see cref="DocumentStatus"/> property to use in sorting.
+        /// <see cref="DocumentFilterProperty"/> for list of properties supported.
+        /// </summary>
+        public DocumentFilterProperty Property { get; set; }
+        /// <summary>
+        /// Convert the order filter to string.
+        /// </summary>
+        /// <returns></returns>
+        public override string ToString()
```
[the remark this answers is not in the document]
Comment (mohamedshabanofficial) [reply sdk/translation/Azure.AI.Translation.Document/src/DocumentFilterOrder.cs:37]: cannot change access modifiers when overriding 'access' inherited member!

File: sdk/translation/Azure.AI.Translation.Document/src/TranslationFilter.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        /// <summary>
+        /// Initializes and instance of <see cref="TranslationFilter"/>.
+        /// </summary>
+        /// <param name="createdAfter">Filter translations by <see cref="TranslationStatus.CreatedOn"/>. Get translations created AFTER a certain datetime.</param>
+        /// <param name="createdBefore">Filter translations by <see cref="TranslationStatus.CreatedOn"/>. Get translations created BEFORE a certain datetime.</param>
+        /// <param name="ids">Filter results by <see cref="TranslationStatus.Id"/>. Get translations with certain IDs.</param>
+        /// <param name="orderBy">Defines sorting options for the result. <see cref="TranslationFilterOrder"/> for more info on which sorting options to use.</param>
+        /// <param name="statuses">Filter results by <see cref="TranslationStatus.Status"/>.</param>
+        public TranslationFilter(
```
[the remark this answers is not in the document]
Comment (mohamedshabanofficial) [reply sdk/translation/Azure.AI.Translation.Document/src/TranslationFilter.cs:23]: exactly!
this is why i made all parameters nullable

File: sdk/translation/Azure.AI.Translation.Document/src/DocumentTranslationClient.cs [hunk 9]
```diff
@@ -226,7 +228,13 @@ Page<TranslationStatus> FirstPageFunc(int? pageSizeHint)
 
                 try
                 {
-                    ResponseWithHeaders<TranslationsStatus, DocumentTranslationGetTranslationsStatusHeaders> response = _serviceRestClient.GetTranslationsStatus(cancellationToken: cancellationToken);
+                    var response = _serviceRestClient.GetTranslationsStatus(
+                        ids: filter.Ids?.Select(id => new Guid(id)),
+                        statuses: filter.Statuses?.Select(status => status.ToString()),
+                        createdDateTimeUtcStart: filter.CreatedAfter,
+                        createdDateTimeUtcEnd: filter.CreatedBefore,
+                        orderBy: filter.OrderBy?.Select(order => order.ToString()),
```
[the remark this answers is not in the document]
Comment (mohamedshabanofficial) [reply sdk/translation/Azure.AI.Translation.Document/src/DocumentTranslationClient.cs]: update code, how about now?

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+            var testIds = operation.GetAllDocumentStatuses().Select(d => d.Id).ToList().GetRange(0, 1);
+
+            // list docs
+            var filter = new DocumentFilter(
+                ids: testIds
+            );
+            var result = operation.GetAllDocumentStatuses(filter: filter);
+
+            // assert
+            Assert.That(result.All(d => testIds.Contains(d.Id)));
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs:68]: looks like `testIds` only has one id, right? will it be better to just do 
`Assert.AreEqual(testIds[0], results.Id)` or something like that

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            // list docs
+            var filter = new DocumentFilter(
+                orderBy: new List<DocumentFilterOrder> {
+                  new DocumentFilterOrder(property: DocumentFilterProperty.CreatedOn, asc: false)
+                }
+            );
+            var filterResults = operation.GetAllDocumentStatuses(filter: filter);
+
+            // assert
+            var timestamp = Recording.UtcNow;
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs:102]: the idea of the test is to make sure the output is organized by `CreatedOn` so no need to check what the current time is. You can instead verify that the order is correct. 
For example, that `result[i].CreatedOn < result[i+1].CreatedOn` and so on

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+                var translationOp = await client.StartTranslationAsync(input);
+                resultIds.Add(translationOp.Id);
+                if (jobTerminalStatus == DocumentTranslationStatus.Succeeded)
+                {
+                    await translationOp.WaitForCompletionAsync();
+                }
+                else if (jobTerminalStatus == DocumentTranslationStatus.Cancelled)
+                {
+                    await translationOp.CancelAsync(default);
+                    Thread.Sleep(3000); // wait for cancel status to propagate!
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs:41]: do we need to do this per job? or just at the end of the function before returning the `resultIds`

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs (hunk 10 above)
Comment (mohamedshabanofficial) [reply sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs:68]: i made it this way to make it more generic

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+            // list translations with filter
+            var cancelledStatusList = new List<DocumentTranslationStatus> {
+                    DocumentTranslationStatus.Cancelled,
+                    DocumentTranslationStatus.Cancelling
+            };
+            var filter = new TranslationFilter(
+                statuses: cancelledStatusList
+            );
+            var filteredTranslations = await client.GetAllTranslationStatusesAsync(filter: filter).ToEnumerableAsync();
+            var filteredIds = new HashSet<string>(filteredTranslations.Select(t => t.Id).ToList());
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs:67]: Ids should be unique already so no need to have a HashSet

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+        [RecordedTest]
+        public async Task GetAllTranslationStatusesFilterByStatusTest()
+        {
+            // create client
+            var client = GetClient();
+
+            // create test jobs
+            await CreateTranslationJobsAsync(client, jobsCount: 5, docsPerJob: 1, jobTerminalStatus: DocumentTranslationStatus.Succeeded);
+            var cancelledIds = await CreateTranslationJobsAsync(client, jobsCount: 2, jobTerminalStatus: DocumentTranslationStatus.Cancelled);
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs:56]: to reduce the recording files, consider just having 1 type of job per action. if we have at least one successful and one cancelled, that should be ok

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+
+        [RecordedTest]
+        public async Task GetAllTranslationStatusesFilterByIdsTest()
+        {
+            // create client
+            var client = GetClient();
+
+            // create test jobs
+            var allIds = await CreateTranslationJobsAsync(client, jobsCount: 10, docsPerJob: 1, jobTerminalStatus: DocumentTranslationStatus.Succeeded);
+            var targetIds = allIds.GetRange(0, 5);
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs:82]: again to reduce recording file sizes, it should be the same if you just sent 2 jobs, and then filter by one id

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+            // list translations with filter
+            var filter = new TranslationFilter(
+                orderBy: new List<TranslationFilterOrder> {
+                  new TranslationFilterOrder(property: TranslationFilterProperty.CreatedOn, asc: false)
+                }
+            );
+            var filteredTranslations = await client.GetAllTranslationStatusesAsync(filter: filter).ToEnumerableAsync();
+
+            // assert
+            var timestamp = Recording.UtcNow;
```
Comment (maririos) [inline sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs:156]: same comment as in Documents

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs (hunk 12 above)
Comment (mohamedshabanofficial) [reply sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs:41]: actually for each job in order to wait for 'cancel' status to propagate

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs (hunk 11 above)
Comment (mohamedshabanofficial) [reply sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs:102]: i'm using it just as a reference point
which will always guaranteed to be greater than the lastly created job

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs (hunk 14 above)
Comment (mohamedshabanofficial) [reply sdk/translation/Azure.AI.Translation.Document/tests/GetAllTranslationsFilterTests.cs:56]: updated jobs count and docs count in order to make the tests run faster

File: sdk/translation/Azure.AI.Translation.Document/api/Azure.AI.Translation.Document.netstandard2.0.cs [hunk 17]
```diff
@@ -1,5 +1,25 @@
 namespace Azure.AI.Translation.Document
 {
+    public partial class DocumentFilter
+    {
+        public DocumentFilter(System.DateTimeOffset createdAfter = default(System.DateTimeOffset), System.DateTimeOffset createdBefore = default(System.DateTimeOffset), System.Collections.Generic.IList<string> ids = null, System.Collections.Generic.IList<Azure.AI.Translation.Document.DocumentFilterOrder> orderBy = null, System.Collections.Generic.IList<Azure.AI.Translation.Document.DocumentTranslationStatus> statuses = null) { }
```
[the remark this answers is not in the document]
Comment (mohamedshabanofficial) [reply sdk/translation/Azure.AI.Translation.Document/api/Azure.AI.Translation.Document.netstandard2.0.cs:5]: i'm confused!
so how exactly would the user set these filters?
for example filter.Statuses is readonly!
should i create a separate empty constructor? (if yes why?!)

[the remark this answers is not in the document]
Comment (maririos) [reply sdk/translation/Azure.AI.Translation.Document/api/Azure.AI.Translation.Document.netstandard2.0.cs:5]: you could do:
```
var filter = new DocumentFilter();
filter.Statuses = { DocumentTranslationStatus.Failed }
```

[the remark this answers is not in the document]
Comment (maririos) [reply sdk/translation/Azure.AI.Translation.Document/api/Azure.AI.Translation.Document.netstandard2.0.cs:5]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/object-and-collection-initializers#object-initializers-with-collection-read-only-property-initialization

[the remark this answers is not in the document]
Comment (maririos) [reply sdk/translation/Azure.AI.Translation.Document/api/Azure.AI.Translation.Document.netstandard2.0.cs:5]: Here is an example for how FR does it: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormRecognizerClient/RecognizeBusinessCardsLiveTests.cs#L497

[the remark this answers is not in the document]
Comment (mohamedshabanofficial) [reply sdk/translation/Azure.AI.Translation.Document/api/Azure.AI.Translation.Document.netstandard2.0.cs:5]: i updated the code, and my tests samples
and i gotta say, it's super counter-intuitive!!

File: sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs (hunk 5 above)
Comment (mohamedshabanofficial) [reply sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs:80]: tried out the second approach
but the sorting option apparently doesn't work
returned docs count = 0

Comment (maririos) [reply sdk/translation/Azure.AI.Translation.Document/tests/GetAllDocumentsFilterTests.cs:80]: can you create an issue for this so after this PR we can add these tests?

**Review by maririos — APPROVED** [review]

**Pull request merged**